utensor_cgen/transformer/ns_transformer.py

```diff
@@ -13,7 +13,175 @@
+    new_end = end
+    for key in allocate_table:
+      if allocate_table[key]['offsetstart'] >= start and allocate_table[key]['offsetend'] <= end:
+        continue
+      elif allocate_table[key]['offsetstart'] <= start and allocate_table[key]['offsetend'] >= start:
+        new_start = allocate_table[key]['offsetstart']
+        if allocate_table[key]['offsetend'] >= end:
+          new_end = max(new_end, allocate_table[key]['offsetend'])
+        else:
+          new_end = max(end, new_end)
+      elif allocate_table[key]['offsetstart'] >= start and allocate_table[key]['offsetend'] >= start:
+        if allocate_table[key]['offsetend'] >= end:
+          new_end = max(new_end, allocate_table[key]['offsetend'])
+        else:
+          new_end = max(end, new_end)
+    return new_start, new_end
+
+  def _query_time_fromallocate_table(self, allocate_table, start, end):
+    time_start = start
+    time_end = end
+    for key in allocate_table:
+      if allocate_table[key]['start'] >= start and allocate_table[key]['end'] <= end:
+        continue
+      elif allocate_table[key]['start'] <= start and allocate_table[key]['end'] >= start:
+        if allocate_table[key]['end'] >= end:
+          time_end = max(time_end, allocate_table[key]['end'])
+        else:
+          time_end = max(end, time_end)
+      elif allocate_table[key]['start'] >= start and allocate_table[key]['end'] >= start:
+        if allocate_table[key]['end'] >= end:
+          time_end = max(time_end, allocate_table[key]['end'])
+        else:
+          time_end = max(end, time_end)
+    return time_start, time_end
+
+  def _query_result(self, allocate_table, op, offset, length, timestart,
+  timeend):
+    ret = list()
+    for key in allocate_table:
+      '''print(key)
+      print(allocate_table[key]['offsetstart'])
+      print(allocate_table[key]['offsetend'])
+      print(allocate_table[key]['start'])
+      print(allocate_table[key]['end'])
+      '''
+      if (allocate_table[key]['offsetstart'] >= offset and allocate_table[key]['offsetstart'] <= offset + length) or (allocate_table[key]['offsetstart'] <= offset and allocate_table[key]['offsetend'] >= offset):
+        if (allocate_table[key]['start'] >= timestart and allocate_table[key]['start'] <= timeend) or (allocate_table[key]['start']
+        <= timestart and allocate_table[key]['end'] >= timeend):
+          ret.append(op)
+
+    return ret
+
+  def _check(self, ugraph, allocate_table, use_def_table, tensor, tensor_offset_start, tensor_offset_end):
+    valid = False
+    timestart = use_def_table[tensor.name]['start']
+    timeend = use_def_table[tensor.name]['end']
+    offset, length = self._query_offset_fromallocate_table(allocate_table, tensor_offset_start, tensor_offset_end)
+    timestart, timeend = self._query_time_fromallocate_table(allocate_table, timestart, timeend)
+    a = self._query_result(allocate_table, tensor, offset, length, timestart, timeend)
+    if len(a) == 0:
+      return True
```

Comment:
```
if len(a) == 0:
    valid = True
return valid
```

I think single return principle should apply to **ANY** language, not only python.